### ISSUE TITLE
Support for --versions to show details version of subcomponents

### ISSUE THREAD

**Issue by mpkocher** [PacificBiosciences/pbcommand#2]
Quiver has a nice model of getting info a subcomponents, pbcommand should adopt this but use '--versions'.

The pbbarser factory constructor should have an optional subcomponents

``` python
subcomponents = [("h5py", "2.4.0"), ("hdf5", "1.8.15")]
p = get_default_contract_parser(TOOL_ID, __version__, desc, driver_exe, task_type, nproc, resource_types, subcomponents=subcomponents)
```

CLI Example output from quiver:

mkocher@login14-biofx01:ui$ variantCaller --version
  GenomicConsensus version: 1.1.0
  ConsensusCore version: 1.0.1
  h5py version: 2.4.0
  hdf5 version: 1.8.12

**Comment by natechols** [comment]
My commit ports the functionality from Quiver but I don't like calling sys.exit() here - makes unit testing difficult.  Not sure if there's a better way or not.

**Comment by mpkocher** [comment]
Thanks for adding subcomponent version support. I think this is good enough to close.

**Comment by pb-cdunn** [comment]
>  but I don't like calling sys.exit() here - makes unit testing difficult. Not sure if there's a better way or not.

Well, the constructor is doing too much work anyway. The argument parser should be injected. The caller could use a Builder.

But I think the root problem is that the parser is using action-callbacks. Is there no way to implement this with declarative argument-parsing, followed by argument-handling?

**Comment by pb-cdunn** [comment]
``` sh
$ variantCaller --version
GenomicConsensus version: 1.1.0
ConsensusCore version: 1.0.1
h5py version: 2.4.0
hdf5 version: 1.8.12
```

Question from the peanut gallery: Why does anyone care about _versions_? Aren't we all on the tip of mainline.

**Comment by mpkocher** [comment]
I am not sure I understand you comment. The PbParser constructor is only assigning values.

``` python
class PbParser(PbParserBase):

    def __init__(self, tool_contract_parser, arg_parser, *parsers):
        # Tool Contract Parser
        self.tool_contract_parser = tool_contract_parser
        # python wrapper parser.
        self.arg_parser = arg_parser
        # add options, so it will show up via --help
        add_base_options_with_emit_tool_contract(self.arg_parser.parser)
        self.other_parsers = parsers
        # for now assume parsers have the same version, id, ...
        tool_id = tool_contract_parser.tool_id
        version = tool_contract_parser.version
        description = tool_contract_parser.description
        super(PbParser, self).__init__(tool_id, version, description)
```

Perhaps the add_base_options could be removed, but the code seems reasonable to me.

The fundamental entry point is really get_default_contract_parser which is just sugar for creating an instance.

**Comment by pb-cdunn** [comment]
I'm talking about the `PyParser` ctor, which is where I noticed an untested typo:
- https://github.com/PacificBiosciences/pbcommand/commit/a14f1ddaa997645e23d4a08d63ddc789efb1fa57#diff-6f09d8fa41461b7420403b09ff871b40L155

**Comment by mpkocher** [comment]
I don't see the typo. It should not be self.subcomponents. 

add_subcomponent_versions_option(self.parser, subcomponents)  adds the --versions help argparse action.

Here's the output from the example app.

``` bash
(pbcommand_test)pbcommand $> python -m 'pbcommand.cli.examples.dev_txt_app' --versions
        pbcommand : 0.1.0
     my_component : 0.1.0
  my_component_id : 1.2.3
```

**Comment by pb-cdunn** [comment]
Ok. I'm probably wrong. But this is what my eyes see:

```
class PyParser(PbParserBase):

    def __init__(self, tool_id, version, description):
        super(PyParser, self).__init__(tool_id, version, description,
            subcomponents=())
        self.parser = argparse.ArgumentParser(version=version,
                                              description=description,
                                              formatter_class=argparse.ArgumentDefaultsHelpFormatter,
                                              add_help=True)
        if subcomponents:
            add_subcomponent_versions_option(self.parser, subcomponents)
```
- https://github.com/PacificBiosciences/pbcommand/blob/a14f1ddaa997645e23d4a08d63ddc789efb1fa57/pbcommand/models/parser.py#L161

Maybe it's corrected on a later commit? Or am I just reading that wrong? Or is `subcomponents` a global?

**Comment by pb-cdunn** [comment]
Oh, I see the problem. Commit a14f1dd is way down the list:

```
* 285dad3 (HEAD, origin/master, origin/HEAD, master) closer to pep8
* 7f54157 Update docs.
*   5439786 Merge branch 'master' of https://github.com/PacificBiosciences/pbcommand
|\
| *   edbdd88 Merge branch 'master' of https://github.com/PacificBiosciences/pbcommand
| |\
| * | f4044fb optional additional tests in TestApp
* | | ab04d2b Update out of date README.
| |/
|/|
* | f299a66 Generalize printing action.
* | 3f01f3b (tag: 0.1.12) Add subcomponents to dev txt app. Tidy up.
|/
* 2a4b9d4 fixed typo
* a14f1dd --versions option to display list of component versions (from GenomicConsensus); see github issue #2
```

The fix is in 2a4b9d4.

Why not use Pull Requests? That's one of the reasons for the GitHub workflow!
